### PR TITLE
E2E testing using KinD

### PR DESCRIPTION
**Pull request by martinkennelly**
Add option to execute E2E tests using KinD

    * Golang code to ensure VFs created during testing remain
      in test network namespace (KinD worker network namespace)

    * Shell script to fetch test tools, execute E2E tests using
      KinD and teardown KinD test environment

Depends on unmerged #97

### CODE REVIEW

**Comment by mrostecki** [comment]
I tried to test this PR (by checking out directly your branch), but unfortunately I'm getting those errors:

https://paste.opensuse.org/85826568

Any idea what I'm doing wrong? I'm running those with `./hack/test-e2e-kind-cluster.sh 0000:81:00.1` and my devices are:

```
# lshw -c network -businfo  
Bus info          Device     Class          Description
=======================================================
pci@0000:02:00.0  eth0       network        Ethernet Controller X710 for 10GbE SFP+
pci@0000:02:00.1  eth1       network        Ethernet Controller X710 for 10GbE SFP+
pci@0000:81:00.0  eth2       network        Ethernet Controller X710 for 10GbE SFP+
pci@0000:81:00.1  eth3       network        Ethernet Controller X710 for 10GbE SFP+
pci@0000:81:00.2  eth4       network        Ethernet Controller X710 for 10GbE SFP+
pci@0000:81:00.3  eth5       network        Ethernet Controller X710 for 10GbE SFP+
pci@0000:81:0a.0  eth13      network        Ethernet Virtual Function 700 Series
pci@0000:81:0a.1  eth12      network        Ethernet Virtual Function 700 Series
pci@0000:81:0a.2  eth11      network        Ethernet Virtual Function 700 Series
pci@0000:81:0a.3  eth10      network        Ethernet Virtual Function 700 Series
```

**Comment by martinkennelly** [comment]
@mrostecki I have a feeling it is issue with KinD cluster or connectivity to the KinD cluster. Catch me on NPWG slack tomorrow and we can have a look. I sent you an email with slack contact details. I have seen this error when the test cases were unable to connect to KinD cluster so I added this command to the script:

```
export KUBECONFIG="${HOME}/.kube/config"
```

**Comment by martinkennelly** [comment]
I am going to rework this patch to remove calling the pf network namespace setting from being called in the shell script to adding it into the test script itself. Thanks for feedback @mrostecki ! Hoping for more!

File: hack/test-e2e-kind-cluster.sh [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+echo "## label KinD's control-plane-node as sriov capable"
+kubectl label node kind-worker feature.node.kubernetes.io/network-sriov.capable=true --overwrite
+echo "## label KinD worker as worker"
+kubectl label node kind-worker node-role.kubernetes.io/worker= --overwrite
+echo "## building PF/VF netns setter"
+go build -o "${root}/hack/pf-vf-netns-set" "${root}/hack/pf-vf-netns-set.go"
+echo "## retrieving netns path from container"
+netnspath="$(docker inspect --format '{{ .NetworkSettings.SandboxKey }}' "${kind_container}")"
+echo "## deploying monitoring of PF/VF network namespace"
+"${root}hack/pf-vf-netns-set" --pfpciaddress "${test_pf_pci_addr}" --netnspath "${netnspath}" &
```
Comment (mrostecki) [inline hack/test-e2e-kind-cluster.sh:94]: I'd suggest doing it with `sudo -E` if the script is launched by an user.

**Review by mrostecki — CHANGES_REQUESTED** [review]

**Comment by martinkennelly** [comment]
@mrostecki can you retest it? See the file ```docs/test-kind.md```

**Review by mrostecki — APPROVED** [review]
Thanks! New changes LGTM and they work without any additional changes

```
Ran 5 of 5 Specs in 531.699 seconds
SUCCESS! -- 5 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestSriovTests (531.70s)
PASS
coverage: [no statements]
ok      github.com/k8snetworkplumbingwg/sriov-network-operator/test/e2e 531.723s        coverage: [no statements]
make: Leaving directory '/home/agabus/sriov-network-operator'
```

**Comment by pliurh** [comment]
@martinkennelly could you rebase this PR.

If I understand it correctly, we still need a machine with an SRIOV NIC to run this test, right?

**Comment by martinkennelly** [comment]
@pliurh Done. I made one change following rebase captured in the head commit https://github.com/k8snetworkplumbingwg/sriov-network-operator/pull/100/commits/f116730387284e09f1bd57f97a42fc3423be29a5 to disable the webhooks explicitly when testing E2E - a change enabled webhooks by default now. I think they should be enabled when we have sufficient tests for them.

**Pull request merged**